File: rvgen/src/riscv/rv32a.rs
// Copyright 2026 Flavien Solt, Quentin Bordier ETH Zurich.
// SPDX-License-Identifier: GPL-3.0-only

//! Port of `riscv/rv32a.py`; function names and argument order are retained.

use super::rvprotoinstrs::*;

pub const RV32A_OPCODE_AMO: i32 = 47;

pub fn rv32a_lrw(aq: bool, rl: bool, rd: i32, rs1: i32, rs2: i32) -> u32 {
    let funct3 = 2;
    let funct7 = 2 << 2 | (aq as i32) << 1 | (rl as i32);
    instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

pub fn rv32a_scw(aq: bool, rl: bool, rd: i32, rs1: i32, rs2: i32) -> u32 {
    let funct3 = 2;
    let funct7 = 3 << 2 | (aq as i32) << 1 | (rl as i32);
    instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

pub fn rv32a_amoswapw(aq: bool, rl: bool, rd: i32, rs1: i32, rs2: i32) -> u32 {
    let funct3 = 2;
    let funct7 = 1 << 2 | (aq as i32) << 1 | (rl as i32);
    instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

pub fn rv32a_amoaddw(aq: bool, rl: bool, rd: i32, rs1: i32, rs2: i32) -> u32 {
    let funct3 = 2;
    let funct7 = (aq as i32) << 1 | (rl as i32);
    instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

pub fn rv32a_amoandw(aq: bool, rl: bool, rd: i32, rs1: i32, rs2: i32) -> u32 {
    let funct3 = 2;
    let funct7 = 12 << 2 | (aq as i32) << 1 | (rl as i32);
    instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

pub fn rv32a_amoorw(aq: bool, rl: bool, rd: i32, rs1: i32, rs2: i32) -> u32 {
    let funct3 = 2;
    let funct7 = 8 << 2 | (aq as i32) << 1 | (rl as i32);
    instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

pub fn rv32a_amoxorw(aq: bool, rl: bool, rd: i32, rs1: i32, rs2: i32) -> u32 {
    let funct3 = 2;
    let funct7 = 4 << 2 | (aq as i32) << 1 | (rl as i32);
    instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

pub fn rv32a_amomaxw(aq: bool, rl: bool, rd: i32, rs1: i32, rs2: i32) -> u32 {
    let funct3 = 2;
    let funct7 = 20 << 2 | (aq as i32) << 1 | (rl as i32);
    instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

pub fn rv32a_amomaxuw(aq: bool, rl: bool, rd: i32, rs1: i32, rs2: i32) -> u32 {
    let funct3 = 2;
    let funct7 = 28 << 2 | (aq as i32) << 1 | (rl as i32);
    instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

pub fn rv32a_amominw(aq: bool, rl: bool, rd: i32, rs1: i32, rs2: i32) -> u32 {
    let funct3 = 2;
    let funct7 = 16 << 2 | (aq as i32) << 1 | (rl as i32);
    instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

pub fn rv32a_amominuw(aq: bool, rl: bool, rd: i32, rs1: i32, rs2: i32) -> u32 {
    let funct3 = 2;
    let funct7 = 24 << 2 | (aq as i32) << 1 | (rl as i32);
    instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

File: go/riscv/rv32a.go
// Copyright 2026 Flavien Solt, Quentin Bordier ETH Zurich.
// SPDX-License-Identifier: GPL-3.0-only

package riscv

// Bit encoders preserve the original argument order and immediate masking.
const RV32A_OPCODE_AMO int32 = 47

func Rv32a_lrw(aq bool, rl bool, rd int32, rs1 int32, rs2 int32) uint32 {
	var funct3 int32 = 2
	var funct7 int32 = 2<<2 | boolInt(aq)<<1 | boolInt(rl)
	return Instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

func Rv32a_scw(aq bool, rl bool, rd int32, rs1 int32, rs2 int32) uint32 {
	var funct3 int32 = 2
	var funct7 int32 = 3<<2 | boolInt(aq)<<1 | boolInt(rl)
	return Instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

func Rv32a_amoswapw(aq bool, rl bool, rd int32, rs1 int32, rs2 int32) uint32 {
	var funct3 int32 = 2
	var funct7 int32 = 1<<2 | boolInt(aq)<<1 | boolInt(rl)
	return Instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

func Rv32a_amoaddw(aq bool, rl bool, rd int32, rs1 int32, rs2 int32) uint32 {
	var funct3 int32 = 2
	var funct7 int32 = boolInt(aq)<<1 | boolInt(rl)
	return Instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

func Rv32a_amoandw(aq bool, rl bool, rd int32, rs1 int32, rs2 int32) uint32 {
	var funct3 int32 = 2
	var funct7 int32 = 12<<2 | boolInt(aq)<<1 | boolInt(rl)
	return Instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

func Rv32a_amoorw(aq bool, rl bool, rd int32, rs1 int32, rs2 int32) uint32 {
	var funct3 int32 = 2
	var funct7 int32 = 8<<2 | boolInt(aq)<<1 | boolInt(rl)
	return Instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

func Rv32a_amoxorw(aq bool, rl bool, rd int32, rs1 int32, rs2 int32) uint32 {
	var funct3 int32 = 2
	var funct7 int32 = 4<<2 | boolInt(aq)<<1 | boolInt(rl)
	return Instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

func Rv32a_amomaxw(aq bool, rl bool, rd int32, rs1 int32, rs2 int32) uint32 {
	var funct3 int32 = 2
	var funct7 int32 = 20<<2 | boolInt(aq)<<1 | boolInt(rl)
	return Instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

func Rv32a_amomaxuw(aq bool, rl bool, rd int32, rs1 int32, rs2 int32) uint32 {
	var funct3 int32 = 2
	var funct7 int32 = 28<<2 | boolInt(aq)<<1 | boolInt(rl)
	return Instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

func Rv32a_amominw(aq bool, rl bool, rd int32, rs1 int32, rs2 int32) uint32 {
	var funct3 int32 = 2
	var funct7 int32 = 16<<2 | boolInt(aq)<<1 | boolInt(rl)
	return Instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}

func Rv32a_amominuw(aq bool, rl bool, rd int32, rs1 int32, rs2 int32) uint32 {
	var funct3 int32 = 2
	var funct7 int32 = 24<<2 | boolInt(aq)<<1 | boolInt(rl)
	return Instruc_rtype(RV32A_OPCODE_AMO, rd, funct3, rs1, rs2, funct7)
}
